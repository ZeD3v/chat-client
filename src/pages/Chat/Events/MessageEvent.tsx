import React from 'react';
import { format } from 'date-fns';

import { User } from '../../../App';
import { ChatEvents } from '../index';

import styles from './styles/MessageEvent.module.scss';

type MessageEventPayload = {
  username: string;
  message: string;
};

type MessageEvent = {
  type: ChatEvents.MESSAGE;
  id: string;
  timestamp: number;
  payload: MessageEventPayload;
};

type MessageEventProps = {
  user: User | null;
  event: MessageEvent;
};

const MessageEvent: React.FC<MessageEventProps> = ({
  user,
  event,
}: MessageEventProps) => {
  const { username, message } = event.payload;

  return (
    <div
      key={`${username}-${event.timestamp}`}
      className={`${styles.message}${
        username === user?.username ? ` ${styles['message--user']}` : ''
      }`}
    >
      {username !== user?.username && (
        <div className={styles['message-user']}>
          <span>{username}</span>
        </div>
      )}
      <div className={styles['message-bubble']}>
        <p>{message}</p>
      </div>
      <div className={styles['message-timestamp']}>
        <span>{format(new Date(event.timestamp), 'dd/MM/yyyy HH:mm:ss')}</span>
      </div>
    </div>
  );
};

export default MessageEvent;
