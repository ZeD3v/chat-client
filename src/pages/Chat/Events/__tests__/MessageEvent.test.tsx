import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import MessageEvent from '../MessageEvent';

import { ChatEvents } from '../../index';

describe('MessageEvent', () => {
  test('renders correctly', () => {
    const { asFragment } = render(
      <MessageEvent
        user={{ username: 'Test user', token: 'test-token' }}
        event={{
          id: 'testid-1',
          timestamp: 1589041780,
          type: ChatEvents.MESSAGE,
          payload: {
            message: 'This is a test message',
            username: 'Test user',
          },
        }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
