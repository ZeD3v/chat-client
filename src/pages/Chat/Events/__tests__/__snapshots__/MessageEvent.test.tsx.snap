// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageEvent renders correctly 1`] = `
<DocumentFragment>
  <div
    class="message message--user"
  >
    <div
      class="message-bubble"
    >
      <p>
        This is a test message
      </p>
    </div>
    <div
      class="message-timestamp"
    >
      <span>
        19/01/1970 12:24:01
      </span>
    </div>
  </div>
</DocumentFragment>
`;
